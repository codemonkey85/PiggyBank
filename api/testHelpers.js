const testHelpers = {
  normalize: (text) => {
    return text.replace(/\s/g, "").toLowerCase();
  },
  mockRouter: () => {
    return {
      get: jest.fn(),
      post: jest.fn(),
      put: jest.fn(),
      delete: jest.fn()
    };
  },
  mockRepo: () => {
    return {
      account: {
        selectAll: jest.fn(),
        insert: jest.fn(),
        update: jest.fn(),
        delete: jest.fn()
      }
    };
  },
  mockRequest: () => {

  },
  mockResult: () => {
    return {
      json: jest.fn()
    };
  }
};

module.exports = testHelpers;