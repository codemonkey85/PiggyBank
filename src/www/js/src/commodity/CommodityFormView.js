"use strict"

import { html, render as renderHtml } from "../../lib/lit-html/lit-html.js"

const template = (d) => html`
  <div class='modal fade' id='modalForm' tabindex='-1' role='dialog' aria-hidden='true'>
    <div class='modal-dialog modal-xl modal-dialog-centered' role='document'>
      <div class='modal-content'>
        <div class='modal-header'>
          <h3 class='modal-title'>Commodity</h3>
        </div>
        <form class='commodity-editor form'>
          <div class='modal-body'>
            <input id='commodityId' type='hidden' name='id' value='${d.id}' />
            <div class='form-row'>
              <div class='form-group col'>
                <label for='name'>Name</label>
                <input id='name' class='form-control' type='text' name='name' value='${d.name}' />
              </div>
              <div class='form-group col'>
                <label for='type'>Type</label>
                <select id='type' class='form-control'>
                  <option value=''>[Choose Commodity Type]</option>
                  <option value='1'>Currency</option>
                  <option value='2'>Investment</option>
                </select>
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='description'>Description</label>
                <input id='description' class='form-control' type='text' name='description' value='${d.description}' />
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='ticker'>Ticker</label>
                <input id='ticker' class='form-control' type='text' name='ticker' value='${d.ticker}' />
              </div>
              <div class='form-group col'>
                <label for='fraction'>Fraction</label>
                <select id='fraction' class='form-control'>
                  <option>[Choose Fraction]</option>
                  <option value='1'>Whole Values (123)</option>
                  <option value='10'>1/10 (123.4)</option>
                  <option value='100'>1/100 (123.45)</option>
                  <option value='1000'>1/1,000 (123.456)</option>
                  <option value='10000'>1/10,000 (123.4567)</option>
                </select>
              </div>
            </div>
          </div>
          <div class='modal-footer'>
            <div class='form-row'>
              <div class='form-group'>
                <button class='save btn btn-success ml-2'><span class='icon-solid-check'></span> Save</button>
                <button class='cancel btn btn-danger ml-2'><span class='icon-close'></span> Cancel</button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  </div>
`

export class CommodityFormView extends Backbone.View {
  preinitialize(args) {
    this.tagName = "div"
    this.className = ""

    this.events = {
      "click .btn.save": "save",
      "click .btn.cancel": "cancel"
    }
  }

  render() {
    renderHtml(template(this.model.attributes), this.el)
    return this
  }

  save(e) {
    e.preventDefault()

    this.model.set({
      id: this.$("#id").val(),
      name: this.$("#name").val(),
      description: this.$("#description").val(),
      type: this.$("#type").val(),
      ticker: this.$("#ticker").val(),
      fraction: Number(this.$("#fraction").val()),
      version: this.$("#version").val()
    })

    const isNew = this.model.isNew()
    this.model.save({}, {
      success: (model) => { this.saved(isNew, model) },
      error: this.saveError
    })
  }

  saved(isNew, model) {
    this.close()
    if (isNew) {
      this.trigger("saved", model)
    }
    else {
      this.trigger("created", model)
    }
  }

  saveError(model, response, opts) {
    alert("error saving commodity: " + response.responseJSON.message)
  }

  cancel(e) {
    e.preventDefault()
    this.close()
  }

  close() {
    this.$("#modalForm").modal("hide")
  }
}