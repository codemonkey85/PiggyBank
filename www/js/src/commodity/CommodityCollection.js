"use strict"

import { CommodityModel } from "./CommodityModel.js"

export class CommodityCollection extends Backbone.Collection {
  preinitialize() {
    this.model = CommodityModel
    this.url = '/api/commodity'
    this.comparator = "name"
  }
}