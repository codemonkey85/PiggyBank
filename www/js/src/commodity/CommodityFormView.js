"use strict"

import { html, render as renderHtml } from "../../lib/lit-html/lit-html.js"

const template = (d) => html`
  <div class='modal fade' id='modalForm' tabindex='-1' role='dialog' aria-hidden='true'>
    <div class='modal-dialog modal-xl modal-dialog-centered' role='document'>
      <div class='modal-content'>
        <div class='modal-header'>
          <h3 class='modal-title'>Commodity</h3>
        </div>
        <form class='commodity-editor form'>
          <div class='modal-body'>
            <input id='commodityId' type='hidden' name='id' value='${d.id}' />
            <div class='form-row'>
              <div class='form-group col'>
                <label for='name'>Name</label>
                <input id='name' class='form-control' type='text' name='name' value='${d.name}' />
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='type'>Type</label>
                <select id='type' class='form-control'>
                  <option value=''>[Choose Commodity Type]</option>
                  <option value='1'>Currency</option>
                  <option value='2'>Investment</option>
                </select>
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='symbol'>Symbol</label>
                <input id='symbol' class='form-control' type='text' name='symbol' value='${d.symbol}' />
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='description'>Description</label>
                <input id='description' class='form-control' type='text' name='description' value='${d.description}' />
              </div>
            </div>
            <div class='form-row'>
              <div class='form-group col'>
                <label for='ticker'>Ticker</label>
                <input id='ticker' class='form-control' type='text' name='ticker' value='${d.ticker}' />
              </div>
            </div>
          </div>
          <div class='modal-footer'>
            <div class='form-row'>
              <div class='form-group'>
                <button class='save btn btn-success ml-2'><span class='icon-solid-check'></span> Save</button>
                <button class='cancel btn btn-danger ml-2'><span class='icon-close'></span> Cancel</button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  </div>
`

export class CommodityFormView extends Backbone.View {
  preinitialize(args) {
    this.tagName = "div"
    this.className = ""

    this.events = {
      "click .btn.save": "save",
      "click .btn.cancel": "cancel"
    }
  }

  render() {
    renderHtml(template(this.model.attributes), this.el)
    return this
  }

  save(e) {
    e.preventDefault()

    this.model.set({
      id: this.$("#id").val(),
      name: this.$("#name").val(),
      type: this.$("#type").val(),
      symbol: this.$("#symbol").val(),
      description: this.$("#description").val(),
      ticker: this.$("#ticker").val(),
      version: this.$("#version").val()
    })

    const isNew = this.model.isNew()
    this.model.save({}, {
      success: (model) => { this.saved(isNew, model) },
      error: this.saveError
    })
  }

  saved(isNew, model) {
    this.close()
    if (isNew) {
      this.trigger("saved", model)
    } 
    else {
      this.trigger("created", model)
    }
  }

  saveError() {
    alert("error saving commodity")
  }

  cancel(e) {
    e.preventDefault()
    this.close()
  }

  close() {
    this.$("#modalForm").modal("hide")
  }
}