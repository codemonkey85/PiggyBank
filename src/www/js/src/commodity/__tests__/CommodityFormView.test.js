import { mock$ } from "../../__tests__/testHelpers"
import { CommodityCollection } from "../CommodityCollection"
import { CommodityFormView } from "../CommodityFormView"

let view
beforeEach(() => {
  view = new CommodityFormView({ model: window.piggybank.commodities.models[0] })
})

test("CommodityFormView has render method", () => {
  const renderedView = view.render()
  expect(view).toBe(renderedView)
  expect(view.el).toMatchSnapshot()
})

test(".close hides the modal and destroys the view", () => {
  const modal = jest.fn()
  view.$ = jest.fn().mockReturnValue({ modal })

  view.close()

  expect(view.$).toHaveBeenCalledWith("#modalForm")
  expect(modal).toHaveBeenCalledWith("hide")
})

test(".cancel prevents default event action and calls .close", () => {
  view.close = jest.fn()
  const preventDefault = jest.fn()

  view.cancel({ preventDefault })

  expect(preventDefault).toHaveBeenCalled()
  expect(view.close).toHaveBeenCalled()
})

test("successful .save updates the model, closes the form and calls .saved", () => {
  view.$ = mock$

  view.saved = jest.fn()
  view.model.save = jest.fn().mockImplementation((attr, opts) => {
    opts.success.call(view.model, 200, null)
  })

  view.render()

  const preventDefault = jest.fn()
  view.save({ preventDefault })
  expect(preventDefault).toHaveBeenCalled()

  // expect(view.$).toHaveBeenCalledWith("#accountName")
  // expect(mockVal).toHaveBeenCalled()
  // expect(view.$).toHaveBeenCalledWith("#isPlaceholder")
  // expect(mockChecked).toHaveBeenCalled()

  expect(view.model.save).toHaveBeenCalled()
})

test("failed .save calls saveError function", () => {
  view.$ = mock$

  view.saveError = jest.fn()
  view.model.save = jest.fn().mockImplementation((attr, opts) => {
    opts.error.call(view.model, 404, null)
  })

  view.render()
  const preventDefault = jest.fn()
  view.save({ preventDefault })

  expect(view.saveError).toHaveBeenCalled()
})

test("saveError alerts user about error", () => {
  const mockAlert = jest.fn()
  window.alert = mockAlert

  view.render()
  view.saveError({}, { responseJSON: { message: "Yikes!" } }, {})

  expect(mockAlert).toHaveBeenCalledWith("error saving commodity: Yikes!")
})

test("saved(isNew=true) closes view and triggers saved event", () => {
  view.close = jest.fn()
  view.trigger = jest.fn()

  view.render()
  const mockModel = {}
  view.saved(true, mockModel)

  expect(view.close).toHaveBeenCalled()
  expect(view.trigger).toHaveBeenCalledWith("saved", mockModel)
})

test("saved(isNew=false) closes view and triggers created event", () => {
  view.close = jest.fn()
  view.trigger = jest.fn()

  view.render()
  const mockModel = {}
  view.saved(false, mockModel)

  expect(view.close).toHaveBeenCalled()
  expect(view.trigger).toHaveBeenCalledWith("created", mockModel)
})