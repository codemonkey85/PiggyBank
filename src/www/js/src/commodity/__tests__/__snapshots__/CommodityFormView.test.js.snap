// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommodityFormView has render method 1`] = `
<div>
  <!---->
  
  
  <div
    aria-hidden="true"
    class="modal fade"
    id="modalForm"
    role="dialog"
    tabindex="-1"
  >
    
    
    <div
      class="modal-dialog modal-xl modal-dialog-centered"
      role="document"
    >
      
      
      <div
        class="modal-content"
      >
        
        
        <div
          class="modal-header"
        >
          
          
          <h3
            class="modal-title"
          >
            Commodity
          </h3>
          
        
        </div>
        
        
        <form
          class="commodity-editor form"
        >
          
          
          <div
            class="modal-body"
          >
            
            
            <input
              id="commodityId"
              name="id"
              type="hidden"
              value="undefined"
            />
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="name"
                >
                  Name
                </label>
                
                
                <input
                  class="form-control"
                  id="name"
                  name="name"
                  type="text"
                  value=""
                />
                
              
              </div>
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="type"
                >
                  Type
                </label>
                
                
                <select
                  class="form-control"
                  id="type"
                >
                  
                  
                  <option
                    value=""
                  >
                    [Choose Commodity Type]
                  </option>
                  
                  
                  <option
                    value="1"
                  >
                    Currency
                  </option>
                  
                  
                  <option
                    value="2"
                  >
                    Investment
                  </option>
                  
                
                </select>
                
              
              </div>
              
            
            </div>
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="description"
                >
                  Description
                </label>
                
                
                <input
                  class="form-control"
                  id="description"
                  name="description"
                  type="text"
                  value=""
                />
                
              
              </div>
              
            
            </div>
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="ticker"
                >
                  Ticker
                </label>
                
                
                <input
                  class="form-control"
                  id="ticker"
                  name="ticker"
                  type="text"
                  value=""
                />
                
              
              </div>
              
              
              <div
                class="form-group col"
              >
                
                
                <label
                  for="fraction"
                >
                  Fraction
                </label>
                
                
                <select
                  class="form-control"
                  id="fraction"
                >
                  
                  
                  <option>
                    [Choose Fraction]
                  </option>
                  
                  
                  <option
                    value="1"
                  >
                    Whole Values (123)
                  </option>
                  
                  
                  <option
                    value="10"
                  >
                    1/10 (123.4)
                  </option>
                  
                  
                  <option
                    value="100"
                  >
                    1/100 (123.45)
                  </option>
                  
                  
                  <option
                    value="1000"
                  >
                    1/1,000 (123.456)
                  </option>
                  
                  
                  <option
                    value="10000"
                  >
                    1/10,000 (123.4567)
                  </option>
                  
                
                </select>
                
              
              </div>
              
            
            </div>
            
          
          </div>
          
          
          <div
            class="modal-footer"
          >
            
            
            <div
              class="form-row"
            >
              
              
              <div
                class="form-group"
              >
                
                
                <button
                  class="save btn btn-success ml-2"
                >
                  <span
                    class="icon-solid-check"
                  />
                   Save
                </button>
                
                
                <button
                  class="cancel btn btn-danger ml-2"
                >
                  <span
                    class="icon-close"
                  />
                   Cancel
                </button>
                
              
              </div>
              
            
            </div>
            
          
          </div>
          
        
        </form>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  
  

  <!---->
</div>
`;
